client/history/caching_redirector.go

```diff
@@ -57,25 +60,46 @@ type (
 		}
 
 		connections            connectionPool
+		goros                  goro.Group
 		historyServiceResolver membership.ServiceResolver
 		logger                 log.Logger
+		membershipUpdateCh     chan *membership.ChangedEvent
+		staleTTL               time.Duration
 	}
 )
 
+const cachingRedirectorListener = "cachingRedirectorListener"
+
 func newCachingRedirector(
 	connections connectionPool,
 	historyServiceResolver membership.ServiceResolver,
 	logger log.Logger,
+	staleTTL time.Duration,
 ) *cachingRedirector {
 	r := &cachingRedirector{
 		connections:            connections,
 		historyServiceResolver: historyServiceResolver,
 		logger:                 logger,
+		membershipUpdateCh:     make(chan *membership.ChangedEvent, 1),
+		staleTTL:               staleTTL,
 	}
 	r.mu.cache = make(map[int32]cacheEntry)
+
+	if r.staleTTL > 0 {
+		r.goros.Go(func(ctx context.Context) error {
+			r.eventLoop(ctx)
+			return nil
+		})
+	}
+
 	return r
 }
 
+func (r *cachingRedirector) stop() {
```

Comment:
this is never called in non-test environments?

Reply:
I don't know of way to accurately call this outside of tests. There's no lifecycle start/stop calls for history clients that I can see, please let me know otherwise.

Reply:
Yeah, I think it's fine. I was just asking to confirm